### ISSUE TITLE
Add a rule to enforce surrounding a potentially `never` function with a `try` block

### ISSUE THREAD

**Issue by SNDST00M** [typescript-eslint/typescript-eslint#3924]
- [x] ~~I have tried restarting my IDE and the issue persists.~~
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Configuration**

```JSON
{
  "rules": {
    "@typescript-eslint/no-uncaught-never": "error"
  }
}
```

```TS
// your repro code case
// valid
try { maybeNever() } catch (e) { console.error(e); return; }
try { console.log(JSON.parse("{")); } catch (e) { console.error(e); return; }
// invalid
maybeNever();
console.log(JSON.parse("{"));
```

(I feel like this would hugely improve the TS ecosystem and get people to mark `@error` functions with `| never`)

**Additional Info**

Feel free to edit this, not sure if this is the correct template.

**Versions**

Not really necessary but:

| package                            | version  |
| ---------------------------------- | -------- |
| `@typescript-eslint/eslint-plugin` | `4.26.0` |
| `@typescript-eslint/parser`        | `4.26.0` |
| `TypeScript`                       | `4.3.5`  |
| `ESLint`                           | `7.27.0` |
| `node`                             | `16.6.2` |

**Comment by bradzacher** [comment]
This rule would have to work for "assertion functions" as well as simple "returns never" functions.

This rule **could** work if it also did not require that you wrap it if you specifically annotate the parent function as returning `never`.

Eg:

```ts
declare function assert(value: unknown): asserts value;

function myFunc1(arg: string | null): string {
  assert(arg); // lint error - must be surrounded by a try/catch
  
  return arg + 'foo';
}
function myFunc2(arg: string | null): string | never {
  assert(arg); // ignored by lint rule because this function is annotated to return `never`

  return arg + 'foo';
}
```

**Comment by SNDST00M** [comment]
Right apparently this feature is more or less "checked exceptions". Which the dev community has a love-hate relationship with.

Though this argument:

> "Checked exceptions are bad because programmers just abuse them by always catching them and dismissing them which leads to problems being hidden and ignored that would otherwise be presented to the user".

Is bullshit imho - users are responsible for abusing features.

**Comment by Gerrit0** [comment]
My objection to this is not because of the "checked exceptions" component, though you are correct that I don't particularly like that either.

I object because this is an abuse of the type system. `string | never` is indistinguishable from `string` - for good reason! This is heavily relied on in type helpers. `T = Exclude<1 | 2 | "a" | "b", number>` is `never | never | "a" | "b"`... which is just `"a" | "b"`. Attaching any meaning to `never` when present in a union is like attaching meaning to expressions containing `+ 0`, yes, you can do it, but it doesn't make sense.

**Comment by SNDST00M** [comment]
Oh that I didn't know. I'll be closing this and perhaps a new issue should be opened using the `@error` tag instead.

**Comment by bradzacher** [comment]
FYI - @Gerrit0 your statement is not entirely correct.
When used in a union - `never` is collapsed into other types in any location that ***IS NOT*** a return type location.

```ts
type T = string | never; // T === string;
type TF = () => string | never; // TF === () => string | never;

function foo(): string | never { // foo === () => string | never
    return '';
}

type TF2 = () => T; // TF2 === () => string;
                    // because the | never was originally declared in non-return type location
```
[repl](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKlC8UDOwBOBLAdgcygHykwgDcJUBuKAeitgXkRQx3IChRJYAxBKACgCUCAHzI0WXASKkK1WjB4NEgkWOaTCJMm1YAzAK6YAxsHQB7TFF1mzggFxqJ+TTKgBvOVZv1lQ+KKYnKS1UVihwqFQIYH1USwByeLYAX1Z2cGgFACZeFX9YShpuHKV+PwDxFjCImtq6zwAjCCMAQ30kaGAAC2hg1wB3FqQoMwxsLBaAG0mQKAATZsmWqLmoLEILAFoomLioDmhJs1bTCyA)

Which means that declaring a function as `() => string | never` ***is entirely valid***.
This rule *does* make sense in TS's type system.

**Comment by Gerrit0** [comment]
What you are seeing there is the language server taking the type node rather than the type. The type `string | never` is *always* `string`, and if you specifically ask the compiler for the type rather than the type node, it will not give you the union.

To prove this, take a look at [ts-ast-viewer](https://ts-ast-viewer.com/#code/C4TwDgpgBAKlC8UDOwBOBLAdgcygHykwgDcJUBuKAeitgXkRQx3IChRJYAxBKACgCUCAHzI0WXASKkK1WjB4NEgkWOaTCJMm1YAzAK6YAxsHQB7TFF1mzggFxqJ+TTKgBvOVZv1lQ+KKYnKS1UVihwqFQIYH1USwByeLYAX1Z2cGgFACZeFX9YShpuHKV+PwDxFjCImtq6zwAjCCMAQ30kaGAAC2hg1wB3FqQoMwxsLBaAG0mQKAATZsmWqLmoLEILAFoomLioDmhJs1bTC1YgA) and click on `foo`, on the right hand side it will give you several pieces of information, but the most important one is the type, which even for your example will give `() => string`

The language server giving you type nodes is it trying to be smart about skipping unnecessary steps. If you route it through the API that uses types (click on the FunctionDeclaration for foo in the middle column, then run `checker.signatureToString(checker.getSignatureFromDeclaration(node))`), you will get `(): string`.

**Comment by bradzacher** [comment]
Huh - I'm wrong yeah.
I didn't dig in and use the APIs - I just took the intellisense at face value.

It's weird that the type printer will fully collapse the type inline (`type T = string | never`), but not with one layer of indirection as a return type (`type T = () => string | never`).
You are correct that when you ask the API for the return type, it does scrub the `never`.

This is definitely impossible then.
